### ISSUE TITLE
Ability to pass additional context to `getTileData`

### ISSUE THREAD

**Issue by supersonicclay** [visgl/deck.gl#4893]
#### Target Use case

Sometimes the implementation of `getTileData` changes, but in a way that only affects getting additional tiles. The current way to inform `TilLayer` that your `getTileData` changed is to pass:

```
new TileLayer({
  getTileData(tile) {
    // do something with tile and somethingYourGetTileDataFuncIsDependentOn
  },
  updateTriggers: {
    getTileData: somethingYourGetTileDataFuncIsDependentOn
  }
})
```

However, when using this approach, `TileLayer` will re-request all the tiles.

#### Proposed feature

Passing the current layer instance to `getTileData` provides a way to give different props to `TileLayer` and have those props passed to `getTileData` for subsequent tile requests without needing to re-request the already-loaded tiles.

With this approach you can accomplish the same like this:

```
new TileLayer({
  getTileData(tile, layer) {
    // do something with tile and layer.props.somethingYourGetTileDataFuncIsDependentOn
  },
  somethingYourGetTileDataFuncIsDependentOn,
})
```

#### To Do List 
  - [ ] Add label and assign to milestone 
  - [ ] Coding
  - [ ] Doc update
  - [ ] What’s new update 
  - [ ] Test

**Comment by kylebarron** [comment]
I still don't entirely understand the case for including `layer`. When would you need such information that you can't already obtain? When you define `getTileData`, you already know which `TileLayer` it's bound to?

**Comment by supersonicclay** [comment]
I'm using react. So, I know the layer instance (as well as the layer props and any component state) on the _first_ render. But for all subsequent renders, the `getTileData` function will be cached. Thus, when `TileLayer` calls `getTileData`, it will use all the variables from the first render's closure. Passing the `layer` allows the `getTileData` to get up-to-date variables (through layer props) from every render. This is particularly relevant with React hooks.

I'll give a more fleshed out example of the issue and what I am proposing. This may be somewhat contrived but hopefully shows the difference. The behavior of `getTileData` changes (by requesting from `b.tile.openstreetmap` instead of `a.tile.openstreetmap`), but in a way that only affects how we get new tiles. We don't need to re-request the existing tiles.

Today:
```js
function MyComponent() {
  const [shard, setShard] = React.useState('a');

  // At some point later, state is modified
  // This causes a re-render of MyComponent and a new TilLayer instance is created (and thus a new `getTileData` function)
  setTimeout(() => setShard('b'), 15000);

  new TileLayer({
    getTileData(tile) {
      const url = `https://${shard}.tile.openstreetmap.org/${z}/${x}/${y}.png`;
      return fetch(url)
          .then((response) => response.blob())
          .then((blob) => createImageBitmap(blob));
    },
    updateTriggers: {
      // Here, we have to tell TileLayer that `getTileData` is dependent on `shard`
      // But this will cause TileLayer to re-query all the tiles it had previously cached
      getTileData: shard
    }
  })
}
```

Proposed:

```js
function MyComponent() {
  const [shard, setShard] = React.useState('a');

  // At some point later, state is modified
  // This causes a re-render of MyComponent and a new TilLayer instance is created (and thus a new `getTileData` function)
  setTimeout(() => setShard('b'), 15000);

  new TileLayer({
    getTileData(tile, layer) {
      const url = `https://${layer.props.shard}.tile.openstreetmap.org/${z}/${x}/${y}.png`;
      return fetch(url)
          .then((response) => response.blob())
          .then((blob) => createImageBitmap(blob));
    },

    // here we can pass state as props so that every layer instance gets up-to-date state
    shard,

  })
}
```

**Comment by supersonicclay** [comment]
Alternatively, if `TileLayer` didn't cache `getTileData`, that would solve my use case also.

**Comment by supersonicclay** [comment]
#4901 fixes it to use the `getTileData` (and other props) handed in to the latest render of `TileLayer` rather than the first render. If we merge #4901, I think we can abandon #4894 if you'd like.

**Comment by Pessimistress** [comment]
Published 8.2.8
